### PR TITLE
[SPARK-48148][CORE] JSON objects should not be modified when read as STRING

### PR DESCRIPTION
**Pull request by eric-maynard**
### What changes were proposed in this pull request?

Currently, when reading a JSON like this:
```
{"a": {"b": -999.99999999999999999999999999999999995}}
```

With the schema:
```
a STRING
```

Spark will yield a result like this:
```
{"b": -1000.0}
```

Other changes such as changes to the input string's whitespace may also occur. In some cases, we apply scientific notation to an input floating-point number when reading it as STRING.

This applies to reading JSON files (as with `spark.read.json`) as well as the SQL expression `from_json`.

### Why are the changes needed?

Correctness issues may occur if a field is read as a STRING and then later parsed (e.g. with `from_json`) after the contents have been modified.

### Does this PR introduce _any_ user-facing change?
Yes, when reading non-string fields from a JSON object using the STRING type, we will now extract the field exactly as it appears.

### How was this patch tested?
Added a test in `JsonSuite.scala`

### Was this patch authored or co-authored using generative AI tooling?
No

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JacksonParser.scala [hunk 1]
```diff
@@ -280,13 +280,32 @@ class JacksonParser(
...
+                case _ =>
+                   // Do nothing in this case; we've already read the token
+              }
+              val endLocation = parser.currentLocation.getByteOffset
+
+              UTF8String.fromBytes(
+                byteArray,
+                startLocation.getByteOffset.toInt,
+                endLocation.toInt - (startLocation.getByteOffset.toInt))
+            case _ =>
```
Comment (eric-maynard) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JacksonParser.scala:302]: It's not clear to me when or if this branch would occur, but I have preserved the existing code here to cover cases where the input may not be readily indexed by a byte offset.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/json/JsonSuite.scala [hunk 2]
```diff
@@ -3865,6 +3865,24 @@ abstract class JsonSuite
       }
     }
   }
+
+  test("SPARK-48148: decimal precision is preserved when object is read as string") {
+    withTempPath { path =>
+
+      val granularFloat = "-999.99999999999999999999999999999999995"
+      val jsonString = s"""{"data": {"v": ${granularFloat}}}, {"data": {"v": ${granularFloat}}}]"""
```
Comment (sadikovi) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/json/JsonSuite.scala:3873]: The JSON string appears to be invalid, for example, it ends with `]` but I don't see any opening bracket for it.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/json/JsonSuite.scala [hunk 3]
```diff
@@ -3865,6 +3865,24 @@ abstract class JsonSuite
       }
     }
   }
+
+  test("SPARK-48148: decimal precision is preserved when object is read as string") {
+    withTempPath { path =>
+
```
Comment (sadikovi) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/json/JsonSuite.scala:3871]: nit: You can remove the new line on L3871.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/json/JsonSuite.scala [hunk 4]
```diff
@@ -3865,6 +3865,24 @@ abstract class JsonSuite
...
+      val jsonString = s"""{"data": {"v": ${granularFloat}}}, {"data": {"v": ${granularFloat}}}]"""
+
+      Seq(jsonString).toDF()
+        .repartition(1)
+        .write
+        .text(path.getAbsolutePath)
+
+      val df = spark.read.schema("data STRING").json(path.getAbsolutePath)
+
+      val expected = s"""{"v": ${granularFloat}}"""
```
Comment (sadikovi) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/json/JsonSuite.scala:3882]: Can you add more test cases for the following? 
- {"data": {"v": "abc"}}, expected: "{"v": "abc"}"
- {"data":{"v": "0.999"}}, expected: "{"v": "0.999"}"
- {"data": [1, 2, 3]}, expected: "[1, 2, 3]"
- {"data": <deeply-nested-object>}, expected the object as string.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JacksonParser.scala [hunk 5]
```diff
@@ -280,13 +280,32 @@ class JacksonParser(
...
           UTF8String.fromString(parser.getText)
 
-        case _ =>
+        case other =>
           // Note that it always tries to convert the data as string without the case of failure.
-          val writer = new ByteArrayOutputStream()
-          Utils.tryWithResource(factory.createGenerator(writer, JsonEncoding.UTF8)) {
-            generator => generator.copyCurrentStructure(parser)
+          val startLocation = parser.getTokenLocation
+          startLocation.contentReference().getRawContent match {
```
Comment (sadikovi) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JacksonParser.scala:286]: Is there an existing API to get the remaining content as string? Also, would it work with multi-line JSON?

**Comment by sadikovi** [comment]
cc @dongjoon-hyun @HyukjinKwon

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JacksonParser.scala (hunk 5 above)
Comment (eric-maynard) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JacksonParser.scala:286]: I was not able to find such an existing API -- there is `JacksonParser.getText` but that appears to simply get the current value if it's a string value.

Comment (eric-maynard) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JacksonParser.scala:286]: wrt. multiline JSON, I have added a test to cover this

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/json/JsonSuite.scala (hunk 4 above)
Comment (eric-maynard) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/json/JsonSuite.scala]: Added more tests -- can you clarify the last example and what we expect that to do? It seems like invalid JSON

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/json/JsonSuite.scala [hunk 6]
```diff
@@ -3865,6 +3865,65 @@ abstract class JsonSuite
...
+              .option("multiLine", multiLine.toString)
+              .json(path.getAbsolutePath)
+            checkAnswer(df, expectedInexactData.map(d => Row(d)))
+          }
+        }
+      }
+      extractData(
+        s"""{"data": {"white":    "space"}}""",
+        expectedInexactData = Seq(s"""{"white":"space"}"""),
+        expectedExactData = Seq(s"""{"white":    "space"}""")
```
Comment (HyukjinKwon) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/json/JsonSuite.scala:3905]:
```suggestion
        expectedExactData = Seq("""{"white":    "space"}""")
```

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/json/JsonSuite.scala [hunk 7]
```diff
@@ -3865,6 +3865,65 @@ abstract class JsonSuite
...
+      )
+      extractData(
+        s"""{"data": ["white",    "space"]}""",
+        expectedInexactData = Seq(s"""["white","space"]"""),
```
Comment (HyukjinKwon) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/json/JsonSuite.scala:3909]:
```suggestion
        expectedInexactData = Seq("""["white","space"]"""),
```

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/json/JsonSuite.scala [hunk 8]
```diff
@@ -3865,6 +3865,65 @@ abstract class JsonSuite
...
+        expectedExactData = Seq(s"""["white",    "space"]""")
```
Comment (HyukjinKwon) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/json/JsonSuite.scala:3910]:
```suggestion
        expectedExactData = Seq("""["white",    "space"]""")
```

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/json/JsonSuite.scala [hunk 9]
```diff
@@ -3865,6 +3865,65 @@ abstract class JsonSuite
...
+      )
+      val granularFloat = "-999.99999999999999999999999999999999995"
+      extractData(
+        s"""{"data": {"v": ${granularFloat}}}""",
+        expectedInexactData = Seq(s"""{"v":-1000.0}"""),
```
Comment (HyukjinKwon) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/json/JsonSuite.scala:3915]:
```suggestion
        expectedInexactData = Seq("""{"v":-1000.0}"""),
```

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/json/JsonSuite.scala [hunk 10]
```diff
@@ -3865,6 +3865,65 @@ abstract class JsonSuite
...
+        expectedExactData = Seq(s"""{"v": ${granularFloat}}""")
+      )
+      // In multiLine, we fall back to the inexact method:
+      extractData(
+        s"""{"data": {"white":\n"space"}}""",
```
Comment (HyukjinKwon) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/json/JsonSuite.scala:3920]:
```suggestion
        """{"data": {"white":\n"space"}}""",
```

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/json/JsonSuite.scala [hunk 11]
```diff
@@ -3865,6 +3865,65 @@ abstract class JsonSuite
...
+        expectedInexactData = Seq(s"""{"white":"space"}"""),
```
Comment (HyukjinKwon) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/json/JsonSuite.scala:3921]:
```suggestion
        expectedInexactData = Seq("""{"white":"space"}"""),
```

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/json/JsonSuite.scala [hunk 12]
```diff
@@ -3865,6 +3865,65 @@ abstract class JsonSuite
...
+        expectedExactData = Seq(s"""{"white":"space"}"""),
```
Comment (HyukjinKwon) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/json/JsonSuite.scala:3922]:
```suggestion
        expectedExactData = Seq("""{"white":"space"}"""),
```

**Review by HyukjinKwon — APPROVED** [review]

**Review by sadikovi — APPROVED** [review]

**Comment by HyukjinKwon** [comment]
```
SPARK-48148: values are unchanged when read as string *** FAILED *** (134 milliseconds)
```

seems it fails

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/json/JsonSuite.scala [hunk 13]
```diff
@@ -3900,26 +3900,26 @@ abstract class JsonSuite
...
       val granularFloat = "-999.99999999999999999999999999999999995"
       extractData(
         s"""{"data": {"v": ${granularFloat}}}""",
-        expectedInexactData = Seq(s"""{"v":-1000.0}"""),
+        expectedInexactData = Seq("""{"v":-1000.0}"""),
         expectedExactData = Seq(s"""{"v": ${granularFloat}}""")
       )
       // In multiLine, we fall back to the inexact method:
       extractData(
-        s"""{"data": {"white":\n"space"}}""",
```
Comment (eric-maynard) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/json/JsonSuite.scala]: This makes `\n` no longer function as a newline here

**Comment by eric-maynard** [comment]
Hey @HyukjinKwon, can you take another look and possible re-trigger tests? I believe multiline should be working now.

**Comment by HyukjinKwon** [comment]
Merged to master.

**Comment by HyukjinKwon** [comment]
btw you can trigger on your own https://github.com/eric-maynard/spark/runs/24789350525 I can't trigger :-).

**Comment by revans2** [comment]
Is there a reason this was only done for scan and not `from_json`/`get_json_object`?  Especially `from_json` where in the past it behaved almost identically to the JSON scan (everything except empty row handling and new line characters).

**Comment by HyukjinKwon** [comment]
@eric-maynard would you mind checking if this is true? we should match the behaviour at least with `from_json`.

**Comment by HyukjinKwon** [comment]
ping @eric-maynard

**Comment by HyukjinKwon** [comment]
sent offline ping too.

**Comment by sandip-db** [comment]
I did some tests and it looks like that this PR doesn't address the same issue with from_json. Json expressions including from_json use [CreateJacksonParser.utf8String](https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/jsonExpressions.scala#L706) to create jackson parser, which in turn converts UTF8String to an InputStreamReader. The latter doesn't allow to copy buffers from random offset. I tried to create jackson parser using the underlying byte array of UTF8String, but ran into [SPARK-16548](https://issues.apache.org/jira/browse/SPARK-16548) (that was fixed by https://github.com/apache/spark/pull/17693).

**Comment by HyukjinKwon** [comment]
Let me take a look

**Comment by HyukjinKwon** [comment]
Hm, I think this PR also doesn't cover multiline cases.

Another problem is that it seems `parser.currentLocation.getByteOffset` doesn't report correctly too so we cannot really get the current byte offset ..

**Comment by HyukjinKwon** [comment]
This is what I tried:

```diff
ff --git a/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/CreateJacksonParser.scala b/sql/catalyst/src/main/scala/org/apache/spark/sql/cat
alyst/json/CreateJacksonParser.scala
index ba7b54fc04e8..721d611bc414 100644
--- a/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/CreateJacksonParser.scala
+++ b/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/CreateJacksonParser.scala
@@ -28,6 +28,10 @@ import org.apache.hadoop.io.Text
 import org.apache.spark.sql.catalyst.InternalRow
 import org.apache.spark.unsafe.types.UTF8String

+private[sql] trait WithByteArrayInputStream {
+  val source: ByteArrayInputStream
+}
+
 object CreateJacksonParser extends Serializable {
   def string(jsonFactory: JsonFactory, record: String): JsonParser = {
     jsonFactory.createParser(record)
@@ -40,7 +44,10 @@ object CreateJacksonParser extends Serializable {
     val bain = new ByteArrayInputStream(
       bb.array(), bb.arrayOffset() + bb.position(), bb.remaining())

-    jsonFactory.createParser(new InputStreamReader(bain, StandardCharsets.UTF_8))
+    jsonFactory.createParser(
+      new InputStreamReader(bain, StandardCharsets.UTF_8) with WithByteArrayInputStream {
+        override val source: ByteArrayInputStream = bain
+      })
   }

   def text(jsonFactory: JsonFactory, record: Text): JsonParser = {
diff --git a/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JacksonParser.scala b/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/
json/JacksonParser.scala
index 32a1731a93d4..d98a14302d5a 100644
--- a/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JacksonParser.scala
+++ b/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JacksonParser.scala
@@ -17,7 +17,7 @@

 package org.apache.spark.sql.catalyst.json

-import java.io.{ByteArrayOutputStream, CharConversionException}
+import java.io.{ByteArrayOutputStream, CharConversionException, InputStreamReader}
 import java.nio.charset.MalformedInputException

 import scala.collection.mutable.ArrayBuffer
@@ -332,6 +332,14 @@ class JacksonParser(
               val buffer = new Array[Byte](size)
               positionedReadable.read(startLocation.getByteOffset, buffer, 0, size)
               UTF8String.fromBytes(buffer, 0, size)
+            case inputStream: InputStreamReader with WithByteArrayInputStream =>
+              skipAhead()
+              val endLocation = parser.currentLocation.getByteOffset
+
+              val size = endLocation.toInt - (startLocation.getByteOffset.toInt)
+              val buffer = new Array[Byte](size)
+              inputStream.source.read(buffer, startLocation.getByteOffset.toInt, size)
+              UTF8String.fromBytes(buffer, 0, size)
             case _ =>
               val writer = new ByteArrayOutputStream()
               Utils.tryWithResource(factory.createGenerator(writer, JsonEncoding.UTF8)) {
diff --git a/sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/JsonExpressionsSuite.scala b/sql/catalyst/src/test/scala/org/apache/spark
/sql/catalyst/expressions/JsonExpressionsSuite.scala
index a23e7f44a48d..12cc8165a39e 100644
--- a/sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/JsonExpressionsSuite.scala
+++ b/sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/JsonExpressionsSuite.scala
@@ -432,6 +432,16 @@ class JsonExpressionsSuite extends SparkFunSuite with ExpressionEvalHelper with
     )
   }

+  test("from_json - nested object to string") {
+    val jsonData = """{"a": {"b": -999.99999999999999999999999999999999995}}"""
+    val schema = StructType(StructField("a", StringType) :: Nil)
+    checkEvaluation(
+      JsonToStructs(schema, Map.empty, Literal(jsonData), UTC_OPT),
+      InternalRow("""{"b": -999.99999999999999999999999999999999995}""")
+    )
+  }
+
+
   test("from_json - invalid data") {
     val jsonData = """{"a" 1}"""
     val schema = StructType(StructField("a", IntegerType) :: Nil)
```

**Pull request closed without merging**